### ISSUE TITLE
Using keras layers within an Estimator either causes training where it shouldn't or corrupts weights

### ISSUE THREAD

**Issue by zmjjmz** [tensorflow/tensorflow#16386]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Debian 3.16.36
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: ('v1.4.0-19-ga52c8d9', '1.4.1')
- **Python version**: 2.7.9
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: See gist

### Describe the problem

Hey there,

So I have a `tf.keras` model that for use on Amazon SageMaker I'm trying to convert into an Estimator. I know there's `tf.keras.estimator.model_to_estimator` but, I'm having separate [issues with that](https://github.com/tensorflow/tensorflow/issues/16385).

In the [easily run reproduction here](https://gist.github.com/zmjjmz/667d0d7e6b6c97c49b3aaf4d67b03d2c) I have (as a demonstration) a `tf.keras.layers.Embedding` which is initialized with all zeros and has `trainable=False`. Followed by that is a `Dense` layer with `use_bias=False` because I couldn't figure out how to get predictions out of an Estimator without training it first (and I can't train nothing apparently). Since all of the embeddings are zero however and can't be trained, the `Dense` layer should always produce a zero, even after training it. Instead, it produces garbage!

In fact, I've taken a few steps to ensure that no training takes place, although ideally I'd be able to just run the estimator without training:
1) I've set the loss to be 0 initially (l2_norm of what should start out as 0)
2) Optimize with SGD using a learning rate of 0
3) One training example that should have zero loss...

The output I actually get is very much non-zero. If I inspect the `embed/embeddings:0` tensor in `tfbdg`, I see this:
```
array([[ 0.14387012,  0.83495581,  0.44025695,  0.25154734,  0.7214781 ,  0.40229702,  0.82108581,  0.12210274,  0.43861651,  0.39615464],                
       [ 0.81636655,  0.48157215,  0.48987687,  0.48775947,  0.62187696,  0.25421095,  0.64555049,  0.97305572,  0.53352964,  0.34286666],                
       [ 0.82881641,  0.80365777,  0.4596678 ,  0.21614265,  0.22256434,  0.07986271,  0.92880177,  0.64946997,  0.89239001,  0.13793337],                
       [ 0.98491704,  0.15281868,  0.77106941,  0.30048406,  0.86042607,  0.88010466,  0.64362776,  0.70185173,  0.49912012,  0.61521161]], dtype=float32)
```

Even though it shouldn't have budged from all zeroes! 

So, something about how I'm doing this is fundamentally broken. I suspect that the issue is in line `61` where I start a new Session -- however this appears to be necessary, since I need to ensure that the `keras` backend is using the same `Graph` as `tensorflow.get_default_graph()` due to the peculiarities of how the Estimator calls the `model_fn`.

Notably those values hold between:
1) Runs of the estimator
2) Successive runs with the same `tf.set_random_seed` value

The latter makes me think that somehow the `Embedding` layer is receiving *a* gradient despite my best efforts, although it's hard to test this versus some sort of memory corruption. 

If you need me to provide any more information let me know -- I'm sure I've left something out.

**Comment by hsm207** [comment]
The solution is to create the embedding layer like this:

```
    emb = keras.layers.Embedding(*(embedding_mat.shape),
                                 input_length=SEQ_LEN, name='embed', trainable=False,
                                 embeddings_initializer='zeros')(inp_placeholder)
```

I think using the `weights` argument to initialize the [Embedding ](https://keras.io/layers/embeddings/)layer has been deprecated.

**Comment by zmjjmz** [comment]
Hm, ok I'll see if I can replicate this using the proper initializer, however the zeros matrix is just for demonstration.

The reason I'm using the `weights` argument here is that in reality I have a bunch of pre-trained embeddings I'm trying to initialize the layer with. Is the correct way to do that now a custom initializer for those embeddings?

**Comment by hsm207** [comment]
If you want to initialize the embedding with some pre-trained weights, then you can pass the [Constant initializer ](https://keras.io/initializers/)to the embedding_initiliazer.

Looking at the [embeddings.py](https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/python/keras/_impl/keras/layers/embeddings.py) source code in TensorFlow, the embeddings are created using the `Layer` class' `add_weight` method which does not accept the `weights` argument. So it really does look like it is deprecated.

**Comment by zmjjmz** [comment]
Ok, I was able to get all zeros by setting the `embedding_initializer=keras.initializers.Zeros()`!

I also didn't realize that the `Constant` initializer would take an arbitrarily shaped value -- I thought it only accepted scalars. It's not super clear in the docs, but I can see that it makes sense now.

Thanks!

I'll close this but it would be nice if there was a deprecation warning that came up when trying to use the `weights` argument.

**Comment by zmjjmz** [comment]
Also, curiously, the use of `weights` is fine if I use a `keras` model and train that rather than an `Estimator`, even if I convert it to an `Estimator` first using the `tf.keras.estimator.model_to_estimator` call.
